Feat(Mesh): Mutual TLS (#4934)
Kong/developer.konghq.com: app/mesh/mutual-tls.md

---
title: 'Configuring Mutual TLS'
description: 'Reference for configuring mutual TLS in {{site.mesh_product_name}}, including CA backends and certificate rotation.'
products:
  - mesh
permalink: /mesh/policies/mutual-tls/
breadcrumbs:
  - /mesh/
content_type: reference
layout: reference
tags:
  - mtls
  - security
  - tls
  - certificates
min_version:
  mesh: '2.9'
---

{:.info}
> If you want to configure version, ciphers or per-service permissive/strict mode, see [`MeshTLS`](/mesh/policies/meshtls).

> The Mutual TLS policy enables automatic encrypted mTLS traffic for all the services in a mesh, and allows you to assign an identity to every [data plane proxy](/mesh/data-plane-proxy/). {{site.mesh_product_name}} supports different types of CA backends as well as automatic certificate rotation.

Reviewer: ```suggestion
The Mutual TLS (mTLS) policy enables automatic encrypted mTLS traffic for all the services in a mesh, and allows you to assign an identity to every [data plane proxy](/mesh/data-plane-proxy/). {{site.mesh_product_name}} supports different types of CA backends as well as automatic certificate rotation.
```

{{site.mesh_product_name}} ships with the following CA (Certificate Authority) supported backends:

* [`builtin`](#using-a-builtin-ca): {{site.mesh_product_name}} automatically generates a CA root certificate and key, which are stored as a [Secret](/mesh/manage-secrets/).
* [`provided`](#using-a-provided-ca): The user provides the CA root certificate and key as a Secret.

> Once you've specified a CA backend, {{site.mesh_product_name}} automatically generates a certificate for every data plane proxy in the mesh. The certificates that {{site.mesh_product_name}} generates are SPIFFE compatible and are used for AuthN/Z use cases in order to identify every workload in the system.

Reviewer: ```suggestion
Once you've specified a CA backend, {{site.mesh_product_name}} automatically generates a certificate for every data plane proxy in the mesh. The certificates that {{site.mesh_product_name}} generates are:

- SPIFFE (
Secure Production Identity Framework for Everyone) compatible.
-  Used for AuthN/Z use cases to identify every workload in the system.
```

The certificates that {{site.mesh_product_name}} generates have a SAN set to `spiffe://<mesh name>/<service name>`. When {{site.mesh_product_name}} enforces policies that require an identity, like [`MeshTrafficPermission`](/mesh/policies/meshtrafficpermission), it will extract the SAN from the client certificate and use it to match the service identity.

{:.warning}

> > By default mTLS **is not** enabled and needs to be explicitly enabled as described below. When mTLS is enabled all traffic is denied **unless** a [`MeshTrafficPermission`](/mesh/policies/meshtrafficpermission) policy is configured to explicitly allow traffic across proxies.

Reviewer: Clarify who needs to do the action:

```suggestion
> By default mTLS **is not** enabled. You need to activate it explicitly as described below. When you enable mTLS, all traffic is denied **unless** a [`MeshTrafficPermission`](/mesh/policies/meshtrafficpermission) you configure a policy to explicitly allow traffic across proxies.
```

> 

> > Always make sure that a [`MeshTrafficPermission`](/mesh/policies/meshtrafficpermission) resource is present before enabling mTLS in a mesh in order to avoid unexpected traffic interruptions caused by a lack of authorization between proxies.

Reviewer: ```suggestion
> Always make sure to include  [`MeshTrafficPermission`](/mesh/policies/meshtrafficpermission) in your configuration before enabling mTLS in a mesh to avoid unexpected traffic interruptions caused by a lack of authorization between proxies.
```

To enable mTLS, configure the `mtls` property in a `Mesh` resource. You can have as many `backends` as you want, but only one at a time can be enabled via the `enabledBackend` property.

> If `enabledBackend` is missing or empty, then mTLS will be disabled for the entire mesh.

Reviewer: ```suggestion
A missing or empty `enabledBackend` property deactivates mTLS for the entire mesh.
```

## Using a builtin CA

A `builtin` CA is the fastest and simplest way to enable mTLS in {{site.mesh_product_name}}.

> With a `builtin` CA backend type, {{site.mesh_product_name}} dynamically generates its own CA root certificate and key that it uses to automatically provision and rotate certificates for every replica of every service.

Reviewer: Break down the info into a list, because it can be long/tricky to get.

```suggestion
With a `builtin` CA backend type, {{site.mesh_product_name}}:

1. Dynamically generates its own CA root certificate and key.
2. Uses the key to automatically provision and rotate certificates for every replica of every service.
```

```

Apply the configuration with `kubectl apply -f [..]`.
{% endnavtab %}

{% navtab "Universal" %}

```yaml
type: Mesh
name: default
mtls:
  enabledBackend: ca-1
  backends:
    - name: ca-1
      type: builtin
      dpCert:
        rotation:
          expiration: 1d
      conf:
        caCert:
          RSAbits: 2048
          expiration: 10y
```

Apply the configuration with `kumactl apply -f [..]` or via the [HTTP API](/mesh/interact-with-control-plane/).
{% endnavtab %}
{% endnavtabs %}

{:.info}
> * The `dpCert` configuration determines how often {{site.mesh_product_name}} should automatically rotate the certificates assigned to every data plane proxy.
> * The `caCert` configuration determines a few properties that {{site.mesh_product_name}} will use when auto-generating the CA root certificate.

### Storing Secrets

When using a `builtin` backend {{site.mesh_product_name}} automatically generates a root CA certificate and key that are stored as a {{site.mesh_product_name}} [Secret resource](/mesh/manage-secrets/) with the following name:

* `<mesh name>.ca-builtin-cert-<backend name>` for the certificate
* `<mesh name>.ca-builtin-key-<backend name>` for the key

> On Kubernetes, {{site.mesh_product_name}} Secrets are stored in the `{{site.mesh_namespace}}` namespace, while on Universal they are stored in the underlying [store](/mesh/control-plane-configuration/#store) configured in `kuma-cp`.

Reviewer: ```suggestion

Secrets are stored depending on the platform:

- Kubernetes stores {{site.mesh_product_name}} Secrets in the `{{site.mesh_namespace}}`.
- Universal stores secrets in the underlying [store](/mesh/control-plane-configuration/#store) configured in `kuma-cp`.
```
